include/ghex/transport_layer/mpi/communicator.hpp

```diff
@@ -28,162 +30,226 @@ namespace gridtools {
-                      * @param args additional arguments to be passed to new container of type Message at construction 
-                      * @return optional which may hold a future< std::tuple<Message,rank_type,tag_type> > */
-                    template<typename Message, typename... Args>
-                    [[nodiscard]] auto recv_any_source(tag_type tag, Args&& ...args) const
+                        GHEX_CHECK_CALLBACK_F(message_type,rank_type,tag_type) 
+                        return recv(message_type{shared_msg_ptr}, src, tag, std::forward<CallBack>(callback));
+                    }
+                    
+                    template<typename Alloc, typename CallBack>
+                    request_cb_type recv(shared_message_buffer<Alloc>& shared_msg, rank_type src, tag_type tag, CallBack&& callback)
                     {
-                        return recv_any<Message>(MPI_ANY_SOURCE, tag, std::forward<Args>(args)...);
-                    }
-
-                    /** @brief non-blocking receive which maches any source and any tag. If a match is found, it
-                      * allocates the container of type Message within this function and returns it in the future.
-                      * The container size will be set according to the matched receive operation.
-                      * @tparam Message a container type
-                      * @tparam Args additional argument types for construction of Message
-                      * @param tag message tag
-                      * @param args additional arguments to be passed to new container of type Message at construction 
-                      * @return optional which may hold a future< std::tuple<Message,rank_type,tag_type> > */
-                    template<typename Message, typename... Args>
-                    [[nodiscard]] auto recv_any_source_any_tag(Args&& ...args) const
+                        GHEX_CHECK_CALLBACK_F(message_type,rank_type,tag_type) 
+                        return recv(message_type{shared_msg.m_message}, src, tag, std::forward<CallBack>(callback));
+                    }
+
+                    template<typename Message, typename CallBack>
+                    lvalue_func<Message&&> recv(Message&& msg, rank_type src, tag_type tag, CallBack&& callback)
                     {
-                        return recv_any<Message>(MPI_ANY_SOURCE, MPI_ANY_TAG, std::forward<Args>(args)...);
+                        GHEX_CHECK_CALLBACK_F(message_type,rank_type,tag_type) 
+                        using V = typename std::remove_reference_t<Message>::value_type;
+                        return recv(message_type{ref_message<V>{msg.data(),msg.size()}}, src, tag, std::forward<CallBack>(callback));
                     }
 
-                private: // implementation
+                    template<typename Message, typename CallBack>
+                    rvalue_func<Message&&> recv(Message&& msg, rank_type src, tag_type tag, CallBack&& callback, std::true_type)
```

Comment:
Wouldn't it make more sense to just call the function with a different type instead of having this extra argument?

Reply:
this extra argument was a bug - I corrected it